File: src/array/lc15.java
package array;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class lc15 {
	public static void main(String[] args) {

	}

	public static List<List<Integer>> threeSum(int[] nums) {
		int size = nums.length;
		List<List<Integer>> ans = new ArrayList<>();
		if (nums == null || size <= 2)
			return ans;
		Arrays.sort(nums); // O(nlogn)
		int left = 1, right = size - 1;
		for (int i = 0; i < size - 2; i++) {
			if (nums[i] > 0)
				break; // 第一个数大于 0，后面的数都比它大，肯定不成立了
			if (i > 0 && nums[i] == nums[i - 1])
				continue; // 去掉重复情况
			int target = -nums[i];
			while (left < right) {
				if (nums[left] + nums[right] == target) {
					ans.add(new ArrayList<>(Arrays.asList(nums[i], nums[left], nums[right])));

					// 现在要增加 left，减小 right，但是不能重复，比如: [-2, -1, -1, -1, 3, 3, 3], i = 0, left = 1,
					// right = 6, [-2, -1, 3] 的答案加入后，需要排除重复的 -1 和 3
					left++;
					right--; // 首先无论如何先要进行加减操作
					while (left < right && nums[left] == nums[left - 1])
						left++;
					while (left < right && nums[right] == nums[right + 1])
						right--;
				} else if (nums[left] + nums[right] < target) {
					left++;
				} else { // nums[left] + nums[right] > target
					right--;
				}
			}
		}
		return ans;

	}
}

File: c++/array/lc15.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

class Solution {
	public:
		vector<vector<int>> threeSum(vector<int>& nums) {
			int len = nums.size();
			vector<vector<int>> res;
			if(len<3)	return res;
			sort(nums.begin(), nums.end());
			int target,x=0,y=len-1;
			for(int i=0; i<len; i++) {
				//ШЅжи
				if(i>0&&nums[i]==nums[i-1])	continue;
				target = nums[i];
				x=i+1;
				y=len-1;
				while(x<y) {
					
//					cout <<i <<" "<<x<<" "<<y<<endl;
					if(x>i+1&&nums[x]==nums[x-1]) {

						++x;
						continue;
					}
					if(y<len-1&&nums[y]==nums[y+1]) {
						--y;continue;
					}
					
					if(nums[x]+nums[y]==-target) {
						vector<int> solve({target,nums[x], nums[y]});
						for (int k = 0; k < 3; k++) {
							cout << solve[k];
						}
						cout << endl;
						res.push_back(solve);
												++x;

					} else if(nums[x]+nums[y]+target<0) {
						++x;
					} else if(nums[x]+nums[y]+target>0) {
						--y;
					}
				}
			}
			return res;
		}
};

int main() {
	vector<int> test({-4,-1,-1,0,1,2}); 
	Solution().threeSum(test);
	return 0;
}
